### ISSUE TITLE
Pipelines paused forever with no error for user if vault credentials do not exist

### ISSUE THREAD

**Issue by keithkroeger** [concourse/concourse#2208]
# Bug Report

We are using VAULT to store secrets for our pipelines.  We have found that if vault tokens are referenced, particularly in resource type definitions, and they don't exist, the user is 'told' that no versions are found and a job is pending when the truth is that the pipeline will never run.

Example:
![image](https://user-images.githubusercontent.com/10414523/39804631-7b34a2ea-536c-11e8-84e0-950a68618e9a.png)

However, if the credentials exist, but are invalid, the user will see information denoting this fact:
![image](https://user-images.githubusercontent.com/10414523/39804644-858f23d2-536c-11e8-9db8-ab6a33ab583c.png)

![image](https://user-images.githubusercontent.com/10414523/39804688-a34d5b0a-536c-11e8-98ae-230993840f15.png)

There are errors listed in the ATC logs, but users certainly can not see these.

* Concourse version: 3.8.0, 3.9.2
* Deployment type (BOSH/Docker/binary): BOSH
* Infrastructure/IaaS: Openstack
* Browser (if applicable): N/A
* Did this used to work? No.

**Comment by colmtuohy** [comment]
+1

**Comment by a459259** [comment]
+1

**Comment by robtec** [comment]
I have also reproduced this issue around the missing vault tokens, on the same environment as @keithkroeger 

Both `vault-image-resource-micky-mouse` and `vault-image-resource-pwd-mouse` don't exist in our Vault

```
---
resource_types:
- name: artifactory-resource
  type: docker-image
  source:
    repository: private.docker.com/springio/artifactory-resource
    username: ((vault-image-resource-micky-mouse))
    password: ((vault-image-resource-pwd-mouse))

resources:
- name: artifactory
  type: artifactory-resource

jobs:
- name: reproduce
  plan:
  - get: artifactory
```

This pipeline results in the `artifactory` resource checking forever :cry: 

![issue-conc](https://user-images.githubusercontent.com/8657653/39871978-71e3930c-545e-11e8-96f1-9a162b61266f.JPG)

Now lets change our pipeline

Instead of using a `get: artifactory` under `jobs.plan`, lets use `put: artifactory`

Bingo, now we get a nice error and the pipeline fails :tada: 

![issue-conc2](https://user-images.githubusercontent.com/8657653/39872065-ab69a350-545e-11e8-9867-1b20c8935eb4.JPG)


Why though?

**Comment by jama-pivotal** [comment]
Throwing a needs-investigation on this as well as putting into the backlog. Sounds like we need some time to dive into the logic to see if we can put the job/resource in an errored state at the point where no credentials are found

**Comment by robtec** [comment]
@jama-pivotal cool cheers, please let us know if you need more information

**Comment by jama-pivotal** [comment]
@robtec 👍 I'm also prioritizing this because I remember a pretty painful debugging session with another user; where the same thing happened but using CredHub. We should try to surface more of these errors when we can, if we can, and where it makes sense

**Comment by topherbullock** [comment]
Dropped in an issue to also expose an API to verify the valid configuration and health of credential managers: #2216

**Comment by robtec** [comment]
@topherbullock thats a great call

**Comment by xtremerui** [comment]
@jama-pivotal @keithkroeger @robtec By running `fly check-resource` on the latest concourse, we can actually see the exact error message you wanted as below:
![screen shot 2018-05-16 at 3 10 16 pm](https://user-images.githubusercontent.com/1585949/40138761-a526990a-591b-11e8-9494-0d4afced8db2.png)

This is done through a recent fix [here](https://github.com/concourse/concourse/issues/1096). Available from concourse 3.12.

Is this good enough for catching the vault misconfiguration error? 

Long term wise, we will need to surface the resource type check error to resource check error so  the error could be shown on resource page correctly i.e. check fail as
![screen shot 2018-05-16 at 5 07 41 pm](https://user-images.githubusercontent.com/1585949/40144176-b1932874-592b-11e8-9691-b7e54acdf9f6.png)

The root cause of the problem is resource check and resource type check are independent. So resource check is waiting for a versioned resource (which could be created only when the resource type is resolved) in DB but it isn't aware of the error from resource type check.

**Comment by robtec** [comment]
@xtremerui Thanks for helping out with this

I've tested my pipeline from above in a `3.13` concourse environment.

```
$ fly -t local check-resource -r test/artifactory
error: check failed with exit status '70':
Expected to find variables: vault-image-resource-micky-mouse
vault-image-resource-pwd-mouse
```

I can confirm we still see the same behavior in the UI, `artifactory` resource checking forever.

Your suggestion on displaying the above error on the resource page would be sweet :tada:

**Comment by jama-pivotal** [comment]
Accepting for now. I think a UX improvement down the road could be highlight that `resource_type` has failed specifically, so users don't get confused

**Comment by keithkroeger** [comment]
I like this as well.

To note, in the logs, the values not being found (in a working credential provider) or a credential provider not being reached result in the same error, Expected to find variables....

Can the logs potentially be more descriptive?

**Comment by pivotal-jwinters** [comment]
We think we've figured out the cause of concourse hanging forever when vault can't login.

Fix is here: 5dd14f62bb24c7691f09c534d0090f4f1f776012

**Comment by keithkroeger** [comment]
Just to be clear, you've introduced a timer to timeout after 5s while getting the creds and you've removed some redundant logic?

A great start!!!
